pkg/completion/completion.go

```diff
@@ -103,7 +99,9 @@ type Completion struct {
 	completed chan struct{}
 
 	// callback is called when Complete is called the first time.
-	callback func()
+	// 'success' tells if the operation was successful or not.
+	// returns updated version number or 0.
+	callback func(success bool) uint64
```

Comment:
This signature is not OK, it's breaking the genericity of this data structure.
I'm ok with passing an argument to the callback. But there's no need to constrain it to a `bool`, so replace `success bool` with `arg interface{}`.

And returning an `uint64` is too specific to your use case. That's a no-no.
I think you can design this without adding this return value.

Reply:
I understand that it is easy for the callback to interpret a generic arguments as it needs to, but the call sites still need to know what the callbacks can accept, no?

Maybe it would be better to add an optional separate callback for the failed completion? Then I would not need to touch the existing callback implementations.

The design intent here is to allow the original completion keep waiting through optional retries. For that to work the version number that the pending completion is waiting for needs to be updated to match the modified version of the resource. Due to the callbacks being called while `AckingResourceMutatorWrapper`'s lock is being held, the callbacks can not call back to the `AckingResourceMutatorWrapper` to Upsert a new version of the resource. In this version of the PR I opted to the owner of the resource (XDSServer) modify the resource itself, without going through `AckingResouerceMutatorWrapper`, and then return the version number of the updated resource. Version numbers are meaningful for the users of completions, but I see that Completion is separated into a package and it is not aware of versions. One alternative would be to launch a goroutine from the NACK callback and call to `AckingResourceMutatorWrapper` from there, and make `AckingResourceMutatorWrapper.Upsert()` check if the passed Completion is already in the list of pending completions. If yes, then update it instead of creating a new pending completion with the same `completion.Completion`. Would this be acceptable?